File: android/app/src/main/java/com/clingsync/android/presentation/Effect.kt
package com.clingsync.android.presentation

import com.clingsync.android.AppSettings

// A side effect the reducer requests and the ViewModel performs. Keeping these
// as data (rather than calling out directly) is what lets the reducers stay pure
// and assertable: a test checks `reduction.effects`, no IO runs.
sealed interface Effect {
    data class EnqueueUpload(val paths: List<String>, val author: String) : Effect

    data object CancelUpload : Effect

    // Verify the repository is really open (the grace close may have dropped
    // it), then either show the browse screen or fall back to a reconnect.
    data object OpenBrowse : Effect

    data class PersistSettings(val settings: AppSettings) : Effect

    // Drop the stored passphrase + encoded URI of a repository we navigated away from.
    data class InvalidateRepository(val repositoryId: String) : Effect

    data object LoadFiles : Effect

    // Open the repository (the ViewModel drives the passphrase/S3 dialogs + gateway).
    data object Connect : Effect

    data object OpenStorageSettings : Effect

    // Return to the main app after a share upload's outcome is acknowledged.
    data object FinishShare : Effect
}

data class Reduction(
    val state: MainUiState,
    val effects: List<Effect> = emptyList(),
)

File: ios/ClingSync/Effect.swift
import Foundation

// A side effect the reducer requests and the store performs. Keeping these as
// data (rather than calling out directly) is what lets the reducers stay pure
// and assertable: a test checks `reduction.effects`, no IO runs.
enum Effect: Equatable {
    case enqueueUpload(ids: [String], author: String)
    case cancelUpload
    case persistSettings(RepositoryConfiguration)
    // Drop the stored passphrase + encoded URI of a repository we left.
    case invalidateRepository(repositoryID: String)
    case loadFiles
    // Open the repository (the store drives the passphrase/S3 prompts + gateway).
    case connect
}

struct Reduction: Equatable {
    var state: AppState
    var effects: [Effect] = []

    static func only(_ state: AppState) -> Reduction {
        Reduction(state: state)
    }
}
